core/math/frustum.cpp

```diff
@@ -0,0 +1,249 @@
+	// we're willing to sacrifice in FOV.
+	real_t add = ((f1 + f2) * (p_oversample - 1.0)) / 2.0;
+	f1 += add;
+	f2 += add;
+	f3 *= p_oversample;
+
+	// always apply KEEP_WIDTH aspect ratio
+	f3 /= p_aspect;
+
+	switch (p_eye) {
+		case 1: { // left eye
+			set_frustum(-f2 * p_z_near, f1 * p_z_near, -f3 * p_z_near, f3 * p_z_near, p_z_near, p_z_far);
+		} break;
+		case 2: { // right eye
+			set_frustum(-f1 * p_z_near, f2 * p_z_near, -f3 * p_z_near, f3 * p_z_near, p_z_near, p_z_far);
+		} break;
+		default: { // mono, does not apply here!
+		} break;
+	}
+}
+
+void Frustum::set_orthogonal(real_t p_left, real_t p_right, real_t p_bottom, real_t p_top, real_t p_znear, real_t p_zfar) {
+	planes[Projection::PLANE_NEAR] = Plane(Vector3(0, 0, 1), -p_znear);
+	planes[Projection::PLANE_FAR] = Plane(Vector3(0, 0, -1), p_zfar);
+	planes[Projection::PLANE_LEFT] = Plane(Vector3(-1, 0, 0), -p_left);
+	planes[Projection::PLANE_TOP] = Plane(Vector3(0, 1, 0), p_top);
+	planes[Projection::PLANE_RIGHT] = Plane(Vector3(1, 0, 0), p_right);
+	planes[Projection::PLANE_BOTTOM] = Plane(Vector3(0, -1, 0), -p_bottom);
+}
+
+void Frustum::set_orthogonal(real_t p_size, real_t p_aspect, real_t p_znear, real_t p_zfar, bool p_flip_fov) {
+	if (!p_flip_fov) {
+		p_size *= p_aspect;
+	}
+
+	set_orthogonal(-p_size / 2, +p_size / 2, -p_size / p_aspect / 2, +p_size / p_aspect / 2, p_znear, p_zfar);
+}
+
+void Frustum::set_frustum(real_t p_size, real_t p_aspect, Vector2 p_offset, real_t p_near, real_t p_far, bool p_flip_fov) {
+	if (!p_flip_fov) {
+		p_size *= p_aspect;
+	}
+
+	set_frustum(-p_size / 2 + p_offset.x, +p_size / 2 + p_offset.x, -p_size / p_aspect / 2 + p_offset.y, +p_size / p_aspect / 2 + p_offset.y, p_near, p_far);
+}
+
+void Frustum::set_frustum(real_t p_left, real_t p_right, real_t p_bottom, real_t p_top, real_t p_near, real_t p_far) {
+	ERR_FAIL_COND(p_right <= p_left);
+	ERR_FAIL_COND(p_top <= p_bottom);
+	ERR_FAIL_COND(p_far <= p_near);
+
+	real_t left = atan2(p_left, p_near);
+	real_t top = atan2(p_top, p_near);
+	real_t right = atan2(p_right, p_near);
+	real_t bottom = atan2(p_bottom, p_near);
+
+	planes[Projection::PLANE_NEAR] = Plane(Vector3(0, 0, 1), -p_near);
+	planes[Projection::PLANE_FAR] = Plane(Vector3(0, 0, -1), p_far);
+	planes[Projection::PLANE_LEFT] = Plane(Vector3(-cos(left), 0, -sin(left)), 0);
+	planes[Projection::PLANE_TOP] = Plane(Vector3(0, cos(top), sin(top)), 0);
+	planes[Projection::PLANE_RIGHT] = Plane(Vector3(cos(right), 0, sin(right)), 0);
+	planes[Projection::PLANE_BOTTOM] = Plane(Vector3(0, -cos(bottom), -sin(bottom)), 0);
+}
+
+Vector<Plane> Frustum::get_projection_planes(const Transform3D &p_transform) const {
+	Vector<Plane> result;
+	Basis tr_inverse = p_transform.basis.inverse().transposed();
+	for (int i = 0; i < 6; i++) {
+		result.push_back(p_transform.xform_fast(planes[i], tr_inverse));
+	}
```

Comment:
I think you could resize `Vector` before iteration (as size is known) and then just directly set values. It may or may not be a little bit faster :)